modules/runner/src/main/java/org/apache/ignite/internal/storage/DistributedConfigurationStorage.java

```diff
@@ -49,56 +82,169 @@ public DistributedConfigurationStorage(MetaStorageManager metaStorageMgr) {
+    @Override public synchronized void addListener(ConfigurationStorageListener listener) {
         listeners.add(listener);
+
+        if (watchId == null) {
+            watchId = metaStorageMgr.registerWatchByPrefix(masterKey, new WatchListener() {
+                @Override public boolean onUpdate(@NotNull Iterable<WatchEvent> events) {
+                    HashMap<String, Serializable> data = new HashMap<>();
+
+                    long maxRevision = 0L;
+
+                    Entry entryForMasterKey = null;
+
+                    for (WatchEvent event : events) {
+                        Entry e = event.newEntry();
+
+                        if (!e.key().equals(masterKey)) {
+                            // TODO: (DISTRIBUTED_PREFIX + ".") should be changed to masterKey.toString().length() when Key from metastorage
+                            // TODO will be replaced with ByteArray https://issues.apache.org/jira/browse/IGNITE-14389
+                            data.put(e.key().toString().substring((DISTRIBUTED_PREFIX + ".").length()),
+                                (Serializable)ByteUtils.fromBytes(e.value()));
+
+                            if (maxRevision < e.revision())
+                                maxRevision = e.revision();
+                        } else
+                            entryForMasterKey = e;
+                    }
+
+                    // Contract of metastorage ensures that all updates of one revision will come in one batch.
+                    // Also masterKey should be updated every time when we update cfg.
+                    // That means that masterKey update must be included in the batch.
+                    assert entryForMasterKey != null;
+
+                    assert maxRevision == entryForMasterKey.revision();
+
+                    assert maxRevision >= version.get();
+
+                    long finalMaxRevision = maxRevision;
+
+                    listeners.forEach(listener -> listener.onEntriesChanged(new Data(data, finalMaxRevision)));
+
+                    return true;
+                }
+
+                @Override public void onError(@NotNull Throwable e) {
+                    LOG.error("Metastorage listener issue", e);
+                }
+            });
+
+        }
     }
 
     /** {@inheritDoc} */
-    @Override public void removeListener(ConfigurationStorageListener listener) {
+    @Override public synchronized void removeListener(ConfigurationStorageListener listener) {
         listeners.remove(listener);
+
+        if (listeners.isEmpty()) {
+            try {
+                metaStorageMgr.unregisterWatch(watchId.get());
+            }
+            catch (InterruptedException | ExecutionException e) {
+                LOG.error("Failed to register watch in metastore", e);
+            }
+
+            watchId = null;
+        }
     }
 
     /** {@inheritDoc} */
     @Override public void notifyApplied(long storageRevision) {
```

Comment:
The semantics of given method is not clear for me. We do not persist `version`, how it'll help to not to repeat the same notification on node restart? Please also pay attention that MetaStorageManager commits every successful big-watch' notification to vault with corresponding appliedRevision and therefore does almost all that we need. Could you please clarify this?

Reply:
let's focus on that method after alpha 2 release, seems that this is needed when restart mechanism will be introduced, added todo with ticket https://issues.apache.org/jira/browse/IGNITE-14697